### PR TITLE
nixos/postgresql: change default superuser

### PR DESCRIPTION
**Pull request by zimbatm**
###### Motivation for this change

Every other distro is using the "postgres" superuser. Tooling shouldn't depend
on a default name for the postgres super user but it tend to do.

This change keeps the old "root" superuser around for existing databases for
backward-compatiblity reasons.

See also #7454, #2097 and #2058
###### Things done
- [ ] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by edolstra** [comment]
The last time we tried to change this it caused a lot of breakage so it was reverted: https://github.com/NixOS/nixpkgs/issues/2097

**Comment by edolstra** [comment]
Would be better to make the user name conditional on `system.stateVersion` (i.e. use `root` if it's less than 16.09, otherwise `postgres`).

**Comment by zimbatm** [comment]
Good idea. It's possible to have both users as superusers right? If yes then we could have both root and postgres as super users, dynamically created in the preStart like you suggested in the previous discussion.

File: nixos/modules/services/databases/postgresql.nix [hunk 1]
```diff
@@ -240,6 +241,14 @@ in
                 sleep 0.1
             done
 
+            # Deprecated - 2016-02-28
+            if ! test -e "${cfg.dataDir}/.new_superuser"; then
+              if ! test -e "${cfg.dataDir}/.old_superuser"; then
+                createuser postgres --superuser
```
Comment (danbst) [inline nixos/modules/services/databases/postgresql.nix]: I have already created `postgres` superuser. Maybe `|| true`?

Comment (zimbatm) [inline nixos/modules/services/databases/postgresql.nix]: Is there a `hasuser` command? `|| true` would be obfuscating other issues.

Comment (danbst) [inline nixos/modules/services/databases/postgresql.nix]: This works for my purposes:
```
      function createUser {
         local user="$1"
         shift
         psql -d postgres -tAc "SELECT 1 FROM pg_roles WHERE rolname='$user'" | grep -q 1 || createuser $user $@
      }
      createUser postgres -s
```

**Comment by zimbatm** [comment]
I the latest change takes all the combinations into account. Let me know if I missed anything and also please test it :)

**Comment by edolstra** [comment]
Sorry, but what I intended was something like:
```
  initdb -U ${if versionOlder config.system.stateVersion "16.09" then "postgres" else "root"}
```
Otherwise, you still have a situation where a system with `stateVersion < 16.09` may or may not have a root user, depending on the previous configurations.

**Comment by danbst** [comment]
> please test it :)

So, when you `initdb -U postgres`, you can no more `psql -d postgres` because user root (who runs `postStart`) is no more created. Probably `psql -U postgres -d postgres` should work, but this will be redundant.

I guess edlostra's solution is the only needed solution, but *requires* to set `system.stateVersion` for fresh deployments.

**Comment by bachp** [comment]
@zimbatm Did you manage to finish this MR? I would be really interested to have this for 17.09

**Comment by zimbatm** [comment]
@bachp unfortunately not, it needs a bit more effort

**Comment by bachp** [comment]
@zimbatm I gave it a try in #25753 let me know if you have some inputs there.

**Pull request closed without merging**